core/src/main/java/io/grpc/Metadata.java

```diff
@@ -362,17 +364,20 @@ private String toStringInternal() {
    *
    * <p>Only the following ASCII characters are allowed in the names of keys:
    * <ul>
-   *   <li>letters, i.e., {@code a-z} and {@code A-Z}</li>
-   *   <li>digits, i.e., {@code 0-9}</li>
-   *   <li>dash, i.e., {@code -}</li>
-   *   <li>underscore, i.e., {@code _}</li>
+   *   <li>letters: {@code a-z} and {@code A-Z}</li>
+   *   <li>digits: {@code 0-9}</li>
+   *   <li>special characters: {@code !#$%&'*+-.^_`|~}</li>
    * </ul>
    *
    * <p>Note this has to be the subset of valid HTTP/2 token characters as defined in RFC7230
-   * Section 3.2.6</p>
+   * Section 3.2.6 and RFC5234 Section B.1</p>
    */
   public abstract static class Key<T> {
 
+    /** Special characters for field names as defined in RFC7230 and RFC5234. */
+    private static final Set<Character> T_CHAR_SPECIAL =
+        ImmutableSet.of('!', '#', '$', '%', '&', '\'', '*', '+', '-', '.', '^', '_', '`', '|', '~');
```

Comment:
could we just add `a`-`z` and `0`-`9`? We'd have to rename to something like ALLOWED_CHARACTERS.